keras/metrics.py

```diff
@@ -47,6 +48,126 @@ def sparse_top_k_categorical_accuracy(y_true, y_pred, k=5):
     return K.mean(K.in_top_k(y_pred, K.cast(K.max(y_true, axis=-1), 'int32'), k), axis=-1)
 
 
+# Stateful Metrics
+
+class Recall(Layer):
+    '''Compute recall over all batches.
+
+    # Arguments
+        name: String, name for the metric.
+        label_encoding: String, 'binary' or 'one_hot', label encoding format.
```

Comment:
What you call `one_hot` is referred to as `categorical` everywhere else in the codebase (e.g. categorical_crossentropy, cateorical_accuracy, to_categorical). But think we don't need this argument, when it comes to single-label classification, because you can automatically infer it from the shape of the predictions (if the last axis is size 1, it's binary, else categorical).

Reply:
But we would need this argument in case we want to support *multi-label* binary classification, where you have multiple dimensions on the last axis, and each one of them encodes a binary class prediction. So I think we should have the argument. And we should add support for:

- multi-class, single-label categorical classification
- multi-class, multi-label binary classification

Reply:
Roger that, I was having issues with the tensor size things. But, I will give that a look.

I really like supporting the multi-class generalization. I will happily add that later today :)

Reply:
Do you want me to add a score threshold argument that would default to 0.5?